File: 1367-linked-list-in-binary-tree/1367-linked-list-in-binary-tree-KMP(Knuth-Morris-Pratt)PatternMatchingAlgorithm.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubPath(self, head: Optional[ListNode], root: Optional[TreeNode]) -> bool:
        # Step 1: Build the pattern (list of values from the linked list)
        # and the prefix table (KMP failure function) for the linked list.
        pattern = [head.val]  # Initialize the pattern with the head node's value
        prefix_table = [0]  # KMP prefix table, starts with 0 for the first element
        pattern_index = 0  # Current index in the pattern

        # Move to the next node in the linked list
        head = head.next

        # Build the pattern array and the KMP prefix table
        while head:
            # If there's a mismatch, move back to the last matching prefix
            while pattern_index > 0 and head.val != pattern[pattern_index]:
                pattern_index = prefix_table[pattern_index - 1]

            # If the current node value matches the pattern, move forward
            pattern_index += 1 if head.val == pattern[pattern_index] else 0

            # Append current value to the pattern and update the prefix table
            pattern.append(head.val)
            prefix_table.append(pattern_index)

            # Move to the next node in the linked list
            head = head.next

        # Step 2: Use DFS to search for the pattern in the binary tree
        return self._search_in_tree(root, 0, pattern, prefix_table)

    def _search_in_tree(self, node: Optional[TreeNode], pattern_index: int, 
                        pattern: List[int], prefix_table: List[int]) -> bool:
        """
        A helper function that performs a DFS to search for the pattern
        in the tree, starting at the current node. Uses the KMP algorithm
        for efficient matching.
        """
        if not node:
            # If we've reached a null node, the pattern cannot be found here
            return False

        # Step 3: Match the current tree node value with the current pattern value
        while pattern_index > 0 and node.val != pattern[pattern_index]:
            # If the current values don't match, move to the previous best match in the pattern
            pattern_index = prefix_table[pattern_index - 1]

        # If the values match, move forward in the pattern
        if node.val == pattern[pattern_index]:
            pattern_index += 1

        # Step 4: Check if we've matched the entire pattern
        if pattern_index == len(pattern):
            return True  # Full pattern match found

        # Step 5: Recursively search both left and right subtrees
        return (
            self._search_in_tree(node.left, pattern_index, pattern, prefix_table) or
            self._search_in_tree(node.right, pattern_index, pattern, prefix_table)
        )

File: Daily Questions/1367-linked-list-in-binary-tree/1367-linked-list-in-binary-tree-KMP(Knuth-Morris-Pratt)PatternMatchingAlgorithm.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

/**
 * Helper function to perform DFS on the binary tree while matching 
 * the linked list pattern using the KMP algorithm.
 * 
 * @param pattern        Array representing the linked list values.
 * @param kmp_table      The KMP table for efficient backtracking.
 * @param pattern_len    Length of the pattern array.
 * @param root           Current binary tree node.
 * @param pattern_index  Current index in the pattern array.
 * @return               True if the pattern is found, otherwise false.
 */
bool searchInTree(int* pattern, int* kmp_table, int pattern_len, 
                  struct TreeNode* root, int pattern_index) {
    // Base case: If we've matched the entire pattern, return true.
    if (pattern_index == pattern_len) {
        return true;
    }
    
    // If the current tree node is NULL, return false (no match).
    if (root == NULL) {
        return false;
    }

    // Adjust the pattern index based on the KMP table for efficient backtracking.
    while (pattern_index >= 0 && pattern[pattern_index] != root->val) {
        pattern_index = kmp_table[pattern_index];
    }

    // Recursively search in both left and right subtrees.
    return searchInTree(pattern, kmp_table, pattern_len, root->left, pattern_index + 1) ||
           searchInTree(pattern, kmp_table, pattern_len, root->right, pattern_index + 1);
}

/**
 * Helper function to build the KMP (Knuth-Morris-Pratt) table.
 * 
 * @param pattern      Array representing the linked list values.
 * @param pattern_len  Length of the pattern array.
 * @return             KMP table for efficient pattern matching.
 */
int* buildKMPTable(int* pattern, int pattern_len) {
    // Allocate memory for the KMP table.
    int* kmp_table = (int*)malloc(pattern_len * sizeof(int));

    // First element of KMP table is always -1.
    kmp_table[0] = -1;
    
    // Handle a single-element pattern.
    if (pattern_len == 1) {
        return kmp_table;
    }

    // Build the KMP table for longer patterns.
    kmp_table[1] = 0;  // Second element is always 0.
    int i = 2, prefix_len = 0;

    // Construct the KMP table.
    while (i < pattern_len) {
        if (pattern[i - 1] == pattern[prefix_len]) {
            kmp_table[i++] = ++prefix_len;
        } else if (prefix_len > 0) {
            prefix_len = kmp_table[prefix_len];
        } else {
            kmp_table[i++] = 0;
        }
    }

    return kmp_table;
}

/**
 * Main function to check if the linked list forms a downward path
 * in the binary tree.
 * 
 * @param head  The head of the linked list.
 * @param root  The root of the binary tree.
 * @return      True if the linked list is a subpath in the binary tree, false otherwise.
 */
bool isSubPath(struct ListNode* head, struct TreeNode* root) {
    // Step 1: Convert the linked list into an array (pattern) and calculate its length.
    int pattern[10000];
    int pattern_len = 0;
    
    struct ListNode* current = head;
    while (current != NULL) {
        pattern[pattern_len++] = current->val;
        current = current->next;
    }

    // Step 2: Build the KMP table for the pattern.
    int* kmp_table = buildKMPTable(pattern, pattern_len);

    // Step 3: Use DFS to search the pattern in the binary tree.
    bool result = searchInTree(pattern, kmp_table, pattern_len, root, 0);

    // Step 4: Free dynamically allocated memory for the KMP table.
    free(kmp_table);
    
    return result;
}
